docs/discussion/Multi-Tenancy Support.md

```diff
@@ -0,0 +1,168 @@
+
+
+## Resource Quota
+Resouce Quota is not as critical as data isolation. In most cases and current implementation, Ratify only has resource quota of CPU/Memory on a pod. However, in certain cases, some tenants may consume too much computing or storage resources which may harm the overall service performance.
+Same as previous topics, the implementation depends on how we deploy Ratify. If Ratify is deployed per namespace, we can configure ResourceQuota and LimitRange to assign limits for Ratify pods per namepsace. On the contrary, a single Ratify deployment cannot utilize ResourceQuota and LimitRange to pre-define the allocation of resources for each namespace, which will offload responsibility to Ratify itself.
+
+
+# Proposed Solutions
+As we already discussed a few areas that needs to be considered for multi-tenancy, there are 2 general options to Ratify's use case. These 2 options are different in terms of how we deploy Ratify service in the cluster.
+
+
+## Ratify Deployment per Namespace or Cluster
+Current Ratify is deployed under gatekeeper-system namespace by default. In the multi-tenancy case, users could either deploy a Ratify service in each namespace or in a specific namespace, such as gatekeeper-system.
+
+Notes: a Ratify service includes all dependency service, such as Redis.
+
+An overview of deployment per namespace:
+
+![](../img/deployment-per-namespace.svg)
+
+An overview of deployment per cluster:
+
+![](../img/deployment-per-cluster.svg)
+
+
+Both options can support multi-tenancy model, each one has its pros and cons, here is a brief comparison of 2 option, and details can be explored in the following sections.
+
+| | Deployment per Namespace | Deployment per Cluster |
+|--|---|---|
+| Core workflow refactoring | easy | hard |
+| Helm templates refactoring | hard | easy |
+| ED request struct | no change | add namespace |
+| ED request number | 1 request per Ratify deployment | 1 request |
+| Resouce Quota | easy | hard |
+| Log isolation | easy | hard |
+| Cache isolation | easy | hard |
+| Resouce efficiency | low | high |
+| Deployment for new namespace | new deployment | no new deployment |
+| mTLS with GK | manually maintain CA certs | no change |
+
+
+### Deployment per Namespace
+#### Pros
+1. Minimal refacotring on Ratify. Since Ratify is deployed to each required namespace, Ratify itself doesn't need to care too much about the data isolation across namespaces. Instead it can offload multi-tenancy responsibility to the control plane.
+2. Distributed cache service can be deployed per namespace as well to support cache isoaltion easily.
+3. No need to add namespace to the external data request from GK to Ratify. We can keep the same request struct.
+4. Users could easily set up ResourceQuota and LimitRange to assign appropriate resources.
+5. As Ratify pods are isolated by namespace, each team can only access its own logs.
+
```

Comment:
I have many concerns on Deployment per Namespace. Users are really care about the resource usage and maintenance effort. That's being said users like light-weight solution and lower-maintenance solution. 

In addition, Ratify as a an addmission controller, users may concern the single-point failure of the Ratify instance. Users will need to maintain multiple Ratify instances in different namespaces. This would be a pain point for users.